gnosis/cowswap
# Add confirmation on approves

# Summary

Adds the confirmation modal when approving a token:

![image](https://user-images.githubusercontent.com/2352112/136347231-64950747-cbb6-4d86-ac12-91f22264d743.png)

This PR is a reiteration on https://github.com/gnosis/cowswap/pull/1483

The confirmation modal should be seen in every case:
- Normal EOAs
- Normal EOAs in expert mode
- Smart contract wallets
- Smart contract wallets in expert mode

 # Notes for the reviewer
- I had to create a mod for `useApproveCallback.ts` so it can now receive a couple of new callbacks which open and close the modal when it's the right time
- I included the modal in the swap page
- I extended `application/actions.ts` enum `ApplicationModal`. It had to be done in the source file, because doing it in the Mod would either break the compatibility of the defined methods, require too much duplication for little gain, or create a dependency loop making it imposible to init the app

 # Not included
- Any design changes. It just use the awesome modal from @biocom 
- Doesn't include it for WRAP and UNWRAP. Although this PR makes it very easy to add it there now. 

 # To Test

1.  Try to approve a token (see tip for testing this easier)
2. Try with expert mode/normal. Try to cancel/reject/.. 


## Tip for testing this easier

![image](https://user-images.githubusercontent.com/2352112/136348073-4438dbc6-975b-4083-b74c-ef9abc9afea1.png)


- Since this workflow requires to start with a token that hasn't been approved
- AND the UI doesn't offer you to un-approve
- You can un-approve WETH by going to: https://rinkeby.etherscan.io/address/0xc778417E063141139Fce010982780140Aa0cD5Ab#writeContract
- Click in "Contract tab", then in "Write contract"
- Connect your wallet
- Input, in the "approve" operation:
   - **guy**: 0xC92E8bdf79f0507f65a392b0ab4667716BFE0110 
   - **wad**: 0
